src/fragment-generation-utils.js

```diff
@@ -965,74 +969,98 @@ const FragmentFactory = class {
+ * Helper class to calculate visible text from the start or end of a range
+ * until a block boundary is reached or the range is exhausted.
  */
 const BlockTextAccumulator = class {
   /**
-   * @param {Range} searchRange - the range for which the text in the last or first non emty block boundary will be calculated 
-   * @param {boolean} isForwardTraversal - true if searchRange if nodes in searchRange will be forward traversed  
+   * @param {Range} searchRange - the range for which the text in the last or
+   *     first non empty block boundary will be calculated
+   * @param {boolean} isForwardTraversal - true if nodes in
+   *     searchRange will be forward traversed
    */
   constructor(searchRange, isForwardTraversal) {
     this.searchRange = searchRange;
-    this.insertFunc = isForwardTraversal ? Array.prototype.push : Array.prototype.unshift;
+    this.isForwardTraversal = isForwardTraversal;
     this.textFound = false;
     this.textNodes = [];
     this.textInBlock = null;
   }
   /**
    * Adds the next node in the search space range traversal to the accumulator.
-   * The accumulator then will keep track of the text nodes in the range until a block boundary is found.
-   * Once a block boundary is found and the content of the text nodes in the boundary is non emtpy, the property
-   * textInBlock will be set with the content of the text nodes, trimmed of leading and trailing whitespaces.
-   * @param {Node} node - next node in the traversal of the searchRange  
+   * The accumulator then will keep track of the text nodes in the range until a
+   * block boundary is found. Once a block boundary is found and the content of
+   * the text nodes in the boundary is non empty, the property textInBlock will
+   * be set with the content of the text nodes, trimmed of leading and trailing
+   * whitespaces.
+   * @param {Node} node - next node in the traversal of the searchRange
    */
   appendNode(node) {
-    // if we already calculated the text in the block boundary just ignore any calls to append nodes
+    // If we already calculated the text in the block boundary just ignore any
+    // calls to append nodes.
     if (this.textInBlock !== null) {
       return;
     }
-    // we found a block boundary, check if there's text inside and set it to textInBlock or keep going to the next block boundary
+    // We found a block boundary, check if there's text inside and set it to
+    // textInBlock or keep going to the next block boundary.
     if (isBlock(node)) {
       if (this.textFound) {
-        // concatenate all the text nodes in the block boundary and trim any trailing and leading whitespaces
-        this.textInBlock = this.textNodes.map(textNode => textNode.textContent).join("").trim();
+        // When traversing backwards the nodes are pushed in reverse order.
+        // Reversing them to get them in the right order.
+        if (!this.isForwardTraversal) {
+          this.textNodes.reverse();
```

Comment:
Yes, it will only be called once because right after this we set textInBlock and in line 1001 we check if textInBlock has value, to skip any further nodes.

Reply:
And we know for sure that textInBlock won't be empty after trimming? Is line 1034 meant to guarantee that?

Reply:
exactly, every time we process a new text node we check if contains something more than whitespaces, which means there's some non empty text in the current block. this guarantees that when we join all the text nodes processed, we'll end up with a non empty string.